geom/type_multi_polygon.go

```diff
@@ -103,101 +121,57 @@ func validateMultiPolygon(polys []Polygon, opts ...ConstructorOption) error {
 	return nil
 }
 
-func polyInteriorsIntersect(p1, p2 Polygon) bool {
-	// Run twice, swapping the order of the polygons each time.
-	for order := 0; order < 2; order++ {
-		p1, p2 = p2, p1
-
-		// Collect points along the boundary of the first polygon. Do this by
-		// first breaking the lines in the ring into multiple segments where
-		// they are intersected by the rings from the other polygon. Collect
-		// the original points in the boundary, plus the intersection points,
-		// then each midpoint between those points. These are enough points
-		// that one of the points will be inside the other polygon iff the
-		// interior of the polygons intersect.
-		allPts := make(map[XY]struct{})
-		for _, r1 := range p1.rings {
-			seq1 := r1.Coordinates()
-			for idx1 := 0; idx1 < seq1.Length(); idx1++ {
-				line1, ok := getLine(seq1, idx1)
-				if !ok {
-					continue
-				}
-				// Collect boundary control points and intersection points.
-				linePts := make(map[XY]struct{})
-				linePts[line1.a] = struct{}{}
-				linePts[line1.b] = struct{}{}
-				for _, r2 := range p2.rings {
-					seq2 := r2.Coordinates()
-					for idx2 := 0; idx2 < seq2.Length(); idx2++ {
-						line2, ok := getLine(seq2, idx2)
-						if !ok {
-							continue
-						}
-						inter := line1.intersectLine(line2)
-						if inter.empty {
-							continue
-						}
-						if inter.ptA != inter.ptB {
-							continue
-						}
-						if inter.ptA != line1.a && inter.ptA != line1.b {
-							linePts[inter.ptA] = struct{}{}
-						}
-					}
-				}
-				// Collect midpoints.
-				if len(linePts) <= 2 {
-					for pt := range linePts {
-						allPts[pt] = struct{}{}
-					}
-				} else {
-					linePtsSlice := make([]XY, 0, len(linePts))
-					for pt := range linePts {
-						linePtsSlice = append(linePtsSlice, pt)
-					}
-					sort.Slice(linePtsSlice, func(i, j int) bool {
-						ptI := linePtsSlice[i]
-						ptJ := linePtsSlice[j]
-						if ptI.X != ptJ.X {
-							return ptI.X < ptJ.X
-						}
-						return ptI.Y < ptJ.Y
-					})
-					allPts[linePtsSlice[0]] = struct{}{}
-					for i := 0; i+1 < len(linePtsSlice); i++ {
-						midpoint := linePtsSlice[i].Midpoint(linePtsSlice[i+1])
-						allPts[midpoint] = struct{}{}
-						allPts[linePtsSlice[i+1]] = struct{}{}
-					}
-				}
+func validatePolyNotInsidePoly(p1, p2 indexedLines) error {
+	// For each point where the boundaries of the two polygons intersect, we
+	// take points clockwise and counterclockwise along the boundaries and see
+	// if they are inside the opposing polygon. If they are, then the interiors
+	// intersect.
+
+	for j := range p2.lines {
+		// Find intersection points.
+		pts := make(map[XY]struct{}) // TODO: Could reuse map
```

Comment:
We don't actually need to use a map here. Instead, we could just store the points in the slice, sort the points, and then do a uniquify pass over the slice.

Reply:
Fixed in 6c638bf